### ISSUE TITLE
Intercept IPv6 ports  

### ISSUE THREAD

**Issue by costinm** [istio/istio#5273]
Not intercepting (or blocking) IPv6 results in direct access to the app port, bypassing envoy.

**Comment by myidpt** [comment]
Update:
`ip6tables` is installed with `iptables`. But the command `ip6tables -t nat -N ISTIO_INBOUND` yields:
```
ip6tables v1.6.0: can't initialize ip6tables table `nat': Table does not exist (do you need to insmod?)
Perhaps ip6tables or your kernel needs to be upgraded.
```
This indicates the kernel module `ip6table_nat` is not enabled. Enable it using `modprobe` does not work, which means the kernel doesn't have the module at all.

Through investigation, the only ways are to install the kernel image or change the container image, either way is pretty messy.

Any better ideas? @costinm

**Comment by costinm** [comment]
You are right, ip6 doesn't have nat.

We'll need to do some research for the equivalent way to intercept incoming.

**Comment by leblancd** [comment]
@myidpt - What kernel version are you using?
I'm not sure if this helps, but we use IPv6 NAT in Kubernetes (supported since Kubernetes version 1.9) for kube-proxy functionality. I expect that it will work for you once you have a kernel that supports it and the required modules. We're running it on various Linux OS/kernels. Currently, I have 2 systems up that use, respectively:
* CentOS 7 with kernel 3.10.0-693.11.6.el7.x86_64
* Ubuntu 16.04 (Alpine image, I believe) with kernel 4.4.0-62-generic
One of the Ubuntu nodes shows:
```
root@kube-node-1:/# lsmod | grep nat
nf_nat_masquerade_ipv6    16384  1 ip6t_MASQUERADE
ip6table_nat           16384  0
nf_nat_ipv6            16384  1 ip6table_nat
ip6_tables             28672  2 ip6table_filter,ip6table_nat
xt_nat                 16384  47
nf_nat_masquerade_ipv4    16384  1 ipt_MASQUERADE
iptable_nat            16384  9
nf_nat_ipv4            16384  1 iptable_nat
ip_tables              24576  3 iptable_filter,iptable_mangle,iptable_nat
x_tables               36864  19 ip6table_filter,xt_statistic,xt_mark,xt_comment,xt_CHECKSUM,xt_recent,ip_tables,xt_tcpudp,ipt_MASQUERADE,ip6t_MASQUERADE,xt_conntrack,xt_nat,iptable_filter,ebtables,ipt_REJECT,iptable_mangle,ip6_tables,xt_addrtype,ip6t_REJECT
nf_nat                 24576  5 nf_nat_ipv4,nf_nat_ipv6,xt_nat,nf_nat_masquerade_ipv4,nf_nat_masquerade_ipv6
nf_conntrack          106496  10 openvswitch,nf_nat,nf_nat_ipv4,nf_nat_ipv6,xt_conntrack,nf_nat_masquerade_ipv4,nf_nat_masquerade_ipv6,nf_conntrack_netlink,nf_conntrack_ipv4,nf_conntrack_ipv6
root@kube-node-1:/# 
```
and the CentOS 7 node shows:
```
[kube@kube-minion-1 ~]$ lsmod | grep nat
nf_nat_masquerade_ipv4    13412  1 ipt_MASQUERADE
iptable_nat            12875  1 
nf_nat_ipv4            14115  1 iptable_nat
nf_nat                 26787  2 nf_nat_ipv4,nf_nat_masquerade_ipv4
nf_conntrack          133387  5 nf_nat,nf_nat_ipv4,xt_conntrack,nf_nat_masquerade_ipv4,nf_conntrack_ipv4
ip_tables              27115  3 iptable_filter,iptable_mangle,iptable_nat
libcrc32c              12644  4 xfs,dm_persistent_data,nf_nat,nf_conntrack
[kube@kube-minion-1 ~]$ 
```

**Comment by pmichali** [comment]
Is IPv6 enabled in sysctl?

**Comment by tiswanso** [comment]
@myidpt -- is the v0.8 comment not supporting IPv6 related to istio itself?  IMO, this specific issue is a GKE/GCP limitation and not an istio issue.

My understanding is GCP/GKE doesn't support IPv6 networks.  You'd have to virtualize the test environment to set it up to avoid the IPv4 only network between GCP hosts.  @leblancd has done that in IPv6 k8s CI environments.

Is the "we will not support IPv6 in v0.8" comment because istio currently has no CI for it?  

I just want to be clear that the "no support" comment isn't leading to doing something in istio that explicitly disables the ipv6 usage at the istio level.

**Comment by leblancd** [comment]
I thought IPv6 support is fairly limited in GKE, e.g. no IPv6 transport between instances? For ingress traffic, I see that GCP supports access from IPv6 clients outside the cluster by proxying the IPv6 packets over IPv4 (perhaps using stateless NAT46 or 6-to-4 transition methods?) to an IPv6 server in the cluster. So maybe this transition method will get the packets to where they'll get IPv6 ingress NAT'ing?

**Comment by myidpt** [comment]
@pmichali IPv6 is enabled on both COS and Ubuntu images.
```
# sysctl net.ipv6.conf.all.disable_ipv6
net.ipv6.conf.all.disable_ipv6 = 0
```
They also have IPv6 addresses in `ifconfig`.

**Comment by pmichali** [comment]
@myidpt I assume forwarding too? (net.ipv6.conf.all.forwarding = 1).  I don't know about GKE or COS, but GCE VM only supports IPv4 (no outside connectivity to IPv6). We had to run a cluster inside of docker containers to be able to get IPv6 support working in Kubernetes (I did the same for Istio in a PoC setup with some hacked code).

**Comment by myidpt** [comment]
@pmichali Forwarding is disabled:
```
# sysctl net.ipv6.conf.all.forwarding
net.ipv6.conf.all.forwarding = 0
```

**Comment by myidpt** [comment]
@tiswanso Istio doesn't support IPv6 nat because the IPv6 nat module may be missing in the node kernel. It's not easy to change GKE node kernel to support IPv6 nat module. We are trying to find out ways to intercept the IPv6 traffic.

Note for the first step, we only reject the "inbound" IPv6 traffic but the "outbound" traffic are still allowed.

Please let me know if this will break existing use cases. We offer a variable `ENABLE_INBOUND_IPV6` to turn off the IPv6 dropping rule.

We could blindly enforce the IPv6 nat rules without checking the returned failures. This will enforce the nat only when there's IPv6 nat module in the system. But this will cause inconsistent behavior (whether supporting IPv6 interception) depending on the node kernel.

**Comment by leblancd** [comment]
@myidpt - Thanks for the explanation!!! This sounds like a good short term approach (reject inbound IPv6 traffic by default, but provide ENABLE_INBOUND_IPV6 for platforms/kernels that support IPv6 NAT). Long term, I'm wondering if adding native IPv6 support in a GKE environment will require more than just installing the right kernel + modules. For example, GCP supports external IPv6 clients via [IPv6 proxy load balancing](https://cloud.google.com/compute/docs/load-balancing/ipv6), but that seems to assume that backends in the cluster are IPv4. Supporting connections from IPv6 external clients to IPv6 internal backends might require translation/tunnels using the IPv4 external IP(s) that is(are) provided in GCP.

**Comment by myidpt** [comment]
@leblancd Thanks for the in-depth info about GCP. We will sync with the GKE team with the IPv6 support.

**Comment by costinm** [comment]
Note this is not only about GCP...

I think it's ok to block ip6 inbound (with option to leave open) if nat table is missing. 
But if nat table exists - we should still do the right thing and intercept it - if not in 0.8 maybe in 0.8.1 or .2

**Comment by myidpt** [comment]
@costinm I agree it's not only about GCP.
I think automatically enabling IPv6 nat based on kernel could cause confusions to users? Maybe use a flag to turn on the IPv6 nat, and by default it's off?

**Comment by hklai** [comment]
@wattli mentioned a short term fix has been merged and this is no longer blocking 0.8.0.

**Comment by costinm** [comment]
Yes, this is no longer blocking 0.8.0 - but we should try to get it soon after ( 0.8.1) 

I don't think we can ship 1.0 with IPv6 disabled - there are clusters that use IPv6, and the fix is not
that big, we need to test if IPv6 and nat tables are enabled, and change few lines of config.

K8S supports IPv6 - and its about 20 years since the first IPv6 draft...

**Comment by eadred** [comment]
With reference to the fix in [#5431](https://github.com/istio/istio/pull/5341), could I suggest that at least loopback IPv6 inbound traffic is allowed, ie include something like `ip6tables -A INPUT -s ::1 -d ::1 -j ACCEPT`?

I am finding in some circumstances that container to container traffic, or even within the same container, requests to `locahost` resolve to ::1 and therefore fail.

As a bit of background I am finding this problem when using the .Net HttpClient in containers based on the .Net Core 2.1 runtime image ([microsoft/dotnet:2.1-aspnetcore-runtime](https://github.com/dotnet/dotnet-docker/blob/master/2.1/aspnetcore-runtime/stretch-slim/amd64/Dockerfile)). I have also seen the same problem when using the Telegraf image ([telegraf:1.5.3-alpine](https://github.com/influxdata/influxdata-docker/blob/9a2e98eb81a6aa5c3e62308b4390e70cd5c1ad67/telegraf/1.5/alpine/Dockerfile)) to collect metrics from a second container in the same pod. Telegraf uses the Go http client.

**Comment by costinm** [comment]
Keeping this open - IPv6 support is pretty important, we're just lacking few small ipv6 ipt6table commands and testing.
